src/map.h

```diff
@@ -3,70 +3,70 @@
  * "LICENSE" for information on usage and redistribution of this file.
  */
 
-/*
- * C Implementation for C++ std::map using red-black tree.
- *
- * Any data type can be stored in a map, just like std::map.
- * A map instance requires the specification of two file types:
- *   1. the key;
- *   2. what data type the tree node will store;
- *
- * It will also require a comparison function to sort the tree.
- */
-
 #pragma once
 
 #include <stdbool.h>
 #include <stddef.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+typedef struct map_node map_node_t;
+struct map_node {
+    void *key;
+    void *val;
+    struct {
+        map_node_t *left, *right_red;
+    } link;
+};
+
+typedef struct rb_tree {
+    map_node_t *root;
+    int (*cmp)(const map_node_t *, const map_node_t *);
+    size_t key_size;
+    size_t val_size;
+} map_internal_t;
+
+typedef map_internal_t *map_t;
+
+typedef struct {
+    map_node_t *prev, *node;
+    size_t count;
+} map_iter_t;
+
+#define map_iter_value(it, type) (*(type *) (it)->node->val)
 
 enum { _CMP_LESS = -1, _CMP_EQUAL = 0, _CMP_GREATER = 1 };
```

Comment:
We shall make use of these enumerations: `_CMP_LESS`, `_CMP_EQUAL`, and `_CMP_GREATER`.